### PR TITLE
accept and verify multiple viewID of view change messages

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: consensus/view_change.go [hunk 1]
```diff
@@ -202,10 +201,13 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
...
 	defer consensus.vcLock.Unlock()
 
+	// update the dictionary key if the viewID is first time received
+	consensus.addViewIDKeyIfNotExist(recvMsg.ViewID)
+
 	// TODO: remove NIL type message
 	// add self m1 or m2 type message signature and bitmap
-	_, ok1 := consensus.nilSigs[consensus.PubKey.SerializeToHexStr()]
-	_, ok2 := consensus.bhpSigs[consensus.PubKey.SerializeToHexStr()]
+	_, ok1 := consensus.nilSigs[recvMsg.ViewID][consensus.PubKey.SerializeToHexStr()]
```
Comment (rlan35) [inline consensus/view_change.go]: Do you need to check both the outer map returns ok and the inner map returns ok? What if there is no map associated with recvMsg.ViewID, any nil pointer issue?

Comment (rlan35) [reply consensus/view_change.go]: I see you have the addViewIDKeyIfNotExist, can you make sure it's called before any potential nil pointer happens?

File: consensus/view_change.go [hunk 2]
```diff
@@ -341,16 +340,17 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
 	utils.Logger().Debug().
 		Str("validatorPubKey", senderKey.SerializeToHexStr()).
 		Msg("[onViewChange] Add M3 (ViewID) type message")
```
Comment (LeoHChen) [inline consensus/view_change.go]: do you need to add viewID into the log messages?  that can help us triage the issues later.

File: consensus/view_change.go [hunk 3]
```diff
@@ -318,13 +318,12 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
 		utils.Logger().Debug().
 			Str("validatorPubKey", senderKey.SerializeToHexStr()).
 			Msg("[onViewChange] Add M1 (prepared) type message")
```
Comment (LeoHChen) [inline consensus/view_change.go]: log message to include viewID as well.

File: consensus/view_change.go (hunk 2 above)
Comment (chaosma) [reply consensus/view_change.go]: yes, I added it.

File: consensus/view_change.go (hunk 3 above)
Comment (chaosma) [reply consensus/view_change.go]: yeap

**Comment by chaosma** [comment]
> Please test the case of a view change happen and then epoch change happen (basically replaying the case of the shard down)

Local test

File: consensus/view_change.go (hunk 1 above)
Comment (chaosma) [reply consensus/view_change.go]: It is called before any access to the map

**Pull request merged**